File: project/ranyoung-onboarding/src/main/java/com/example/ranyoungonboarding/controller/SurveyController.java
package com.example.ranyoungonboarding.controller;

import com.example.ranyoungonboarding.api.CreateSurveyRequest;
import com.example.ranyoungonboarding.api.SubmitResponseRequest;
import com.example.ranyoungonboarding.api.SurveyResponse;
import com.example.ranyoungonboarding.api.ResponseDto;
import com.example.ranyoungonboarding.domain.Response;
import com.example.ranyoungonboarding.domain.Survey;
import com.example.ranyoungonboarding.service.ResponseService;
import com.example.ranyoungonboarding.service.SurveyService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/surveys")
@RequiredArgsConstructor
public class SurveyController {

    private final SurveyService surveyService;
    private final ResponseService responseService;

    // 1. 설문조사 생성 API
    @PostMapping
    public ResponseEntity<SurveyResponse> createSurvey(@Valid @RequestBody CreateSurveyRequest request) {
        Survey survey = surveyService.createSurvey(request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(SurveyResponse.fromEntity(survey));
    }

    // 2. 설문조사 수정 API
    @PutMapping("/{surveyId}")
    public ResponseEntity<SurveyResponse> updateSurvey(
            @PathVariable Long surveyId,
            @Valid @RequestBody CreateSurveyRequest request) {
        Survey survey = surveyService.updateSurvey(surveyId, request);
        return ResponseEntity.ok(SurveyResponse.fromEntity(survey));
    }

    // 3. 설문조사 응답 제출 API
    @PostMapping("/{surveyId}/responses")
    public ResponseEntity<ResponseDto> submitResponse(
            @PathVariable Long surveyId,
            @Valid @RequestBody SubmitResponseRequest request) {
        Response response = responseService.submitResponse(surveyId, request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ResponseDto.fromEntity(response));
    }

    // 4. 설문조사 응답 조회 API
    @GetMapping("/{surveyId}/responses")
    public ResponseEntity<Page<ResponseDto>> getSurveyResponses(
            @PathVariable Long surveyId,
            @RequestParam(required = false) String questionName,
            @RequestParam(required = false) String answerValue,
            Pageable pageable) {

        Page<Response> responses;
        if (questionName != null && answerValue != null) {
            // Advanced: 검색 기능 사용
            responses = responseService.searchSurveyResponses(surveyId, questionName, answerValue, pageable);
        } else {
            // 기본: 전체 응답 조회
            responses = responseService.getSurveyResponses(surveyId, pageable);
        }

        return ResponseEntity.ok(responses.map(ResponseDto::fromEntity));
    }

    // 추가: 설문조사 조회 API
    @GetMapping("/{surveyId}")
    public ResponseEntity<SurveyResponse> getSurvey(@PathVariable Long surveyId) {
        Survey survey = surveyService.getSurvey(surveyId);
        return ResponseEntity.ok(SurveyResponse.fromEntity(survey));
    }
}

File: project/doohyeon-onboarding/src/main/kotlin/com/example/doohyeononboarding/api/controller/SurveyController.kt
package com.example.doohyeononboarding.api.controller

import com.example.doohyeononboarding.api.service.SurveyService
import com.example.doohyeononboarding.api.service.request.CreateSurveyRequest
import com.example.doohyeononboarding.api.service.request.UpdateSurveyRequest
import com.example.doohyeononboarding.api.service.response.CreateSurveyResponse
import com.example.doohyeononboarding.api.service.response.UpdateSurveyResponse
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class SurveyController(
    private val surveyService: SurveyService,
) {

    /**
     * 설문 등록 API
     *
     * @param request 설문 등록 요청 정보
     *          - title: 설문 제목
     *          - description: 설문 설명
     *          - questions: 설문 항목 리스트
     *              - title: 설문 항목 제목
     *              - description: 설문 항목 설명
     *              - type: 설문 항목 유형 (단답형, 장문형, 단일 선택, 다중 선택)
     *              - isRequired: 항목 필수 여부
     *              - options: 선택형일 경우 선택지 리스트
     *
     * @return 생성된 설문 ID
     */
    @PostMapping("/survey")
    fun createSurvey(@RequestBody request: CreateSurveyRequest): ResponseEntity<CreateSurveyResponse> {
        val response = surveyService.createSurvey(request)
        return ResponseEntity.ok(response)
    }

    /**
     * 설문 수정 API
     *
     * @param request 설문 수정 요청 정보
     *          - title: 설문 제목
     *          - description: 설문 설명
     *          - questions: 설문 항목 리스트
     *              - title: 설문 항목 제목
     *              - description: 설문 항목 설명
     *              - type: 설문 항목 유형 (단답형, 장문형, 단일 선택, 다중 선택)
     *              - isRequired: 항목 필수 여부
     *              - options: 선택형일 경우 선택지 리스트
     *
     * @return 수정된 설문 ID
     */
    @PatchMapping("/survey/{surveyId}")
    fun updateSurvey(
        @PathVariable surveyId: Long,
        @RequestBody request: UpdateSurveyRequest
    ): ResponseEntity<UpdateSurveyResponse> {
        val response = surveyService.updateSurvey(surveyId, request)
        return ResponseEntity.ok(response)
    }

}
